src/Internal/Int64.elm

```diff
@@ -0,0 +1,144 @@
+module Internal.Int64 exposing (Int64, fromInts, operations, toInts)
+
+import Bitwise
+import Internal.IntOperations exposing (IntOperations)
+
+
+type Int64
+    = Int64 Ints
+
+
+type alias Ints =
+    { lower : Int, higher : Int }
+
+
+fromInts : Int -> Int -> Int64
+fromInts higher lower =
+    Int64 { higher = Bitwise.or 0 higher, lower = Bitwise.or 0 lower }
+
+
+toInts : Int64 -> ( Int, Int )
+toInts (Int64 { higher, lower }) =
+    ( higher, lower )
+
+
+operations : IntOperations Int64
+operations =
+    { toZigZag = toZigZag
+    , fromZigZag = fromZigZag
+    , fromUnsigned = identity
+    , toUnsigned = identity
+    , popBase128 = popBase128
+    , pushBase128 = pushBase128
+    , fromBase128 = fromBase128
+    }
+
+
+toZigZag : Int64 -> Int64
+toZigZag value =
+    xor (shiftRightBy63 value)
+        (shiftLeftBy 1 value)
+
+
+fromZigZag : Int64 -> Int64
+fromZigZag value =
+    xor (shiftRightZfBy 1 value) (value |> andInt 1 |> negate)
+
+
+popBase128 : Int64 -> ( Int, Int64 )
+popBase128 ((Int64 { lower }) as int64) =
+    let
+        base128 =
+            Bitwise.and 0x7F lower
+
+        higherBits =
+            shiftRightZfBy 7 int64
+    in
+    ( base128, higherBits )
+
+
+pushBase128 : Int -> Int64 -> Int64
+pushBase128 sevenBitInt acc =
+    addUnsafe sevenBitInt (shiftLeftBy 7 acc)
+
+
+shiftRightZfBy : Int -> Int64 -> Int64
+shiftRightZfBy n (Int64 { lower, higher }) =
+    if n > 32 then
+        fromInts 0 (Bitwise.shiftRightZfBy n higher)
+
+    else
+        let
+            carry =
+                Bitwise.shiftLeftBy (32 - n) higher
+
+            newLower =
+                lower
+                    |> Bitwise.shiftRightZfBy n
+                    |> Bitwise.or carry
+                    |> Bitwise.shiftRightZfBy 0
+        in
+        fromInts (Bitwise.shiftRightZfBy n higher) newLower
+
+
+shiftRightBy63 : Int64 -> Int64
+shiftRightBy63 (Int64 { higher }) =
+    fromInts
```

Comment:
In the end this either returns `1111...111` or `0000...000`, right?

Reply:
Thanks for noticing. The two operations used there do the same thing.
Since bitwise shifts should be faster than comparison + if/else, we can replace the implementation with
```
let
        zeroOrMinusOne =
            Bitwise.shiftRightBy 31 higher
in
fromInts zeroOrMinusOne zeroOrMinusOne
```